### DISCUSSION TITLE
Wie können wir die Kampagne besser versionieren?

### DISCUSSION THREAD

**Discussion by Naxos84** [Lector/DSA5maptool#3] [Ideas]
Die resultierende Kampagnen Datei ist ja auch nur ein "zip-File".
Wäre es dann vielleicht besser, wenn wir den Inhalt in github als Versionsverwaltung packen, damit man sehen kann was/wann/von wem geändert wurde?
Und das "fertige" Kampagnen File dann unter releases bereit stellt?

Dazu müssten die Dateien die dann hier in der Versionskontrolle liegen theoretisch einfach nur gezippt werden und das "zip-File" in ".cmpgn" umbenannt werden.

**Comment by Lector** [comment]
Das wäre schonmal ein guter Vorschlag.

Ich will trotzdem mal eine Übersicht geben um das ganze Problem etwas detailierter zu beleuchten.
Die *.campaing-File ist wie gesagt "nur" eine Zip-Datei, die man direkt mit maptool direkt öffnen kann. Zumindest der Release sollte auch weiter in diesem Format angeboten werden.
Innhalt dieser Zip-Datei sind:
- ein Asset-Ordner mit allen Bildern. Diese Bilder sind zufällig benannt. Die Namen der Bilder sind IDs auf die dann referenziert wird. Zu jedem Bild existiert noch eine endungslose XML-Datei mit Metainfos wie z.B. dem Dateiformat
- eine properties.xml mit wenigen Metainfos, in Form von Versionsnummern der Kampagne
- Der content.xml. Das ist das "Herzstück" des Frameworks. Darin enthalten sind alle Karten, Kampagneneinstellungen (Zustände, Lichter...) und natürlich Tokens. Da alle Makros auf den 4 Kampagnen-Tokens liegen ist hier auch der Code zu finden. Wobei noch das Problem hinzukommt dass der Code in dieser xml-Datei natürlich auch XML-Codiert ist. So werden Anführungszeichen z.B. als &quot; codiert. Sämtliche Makros sollten als <entry> innerhalb einer <macroPropertiesMap> innerhalb eines Token-Tags aufgelistet sein.

Während die ersten beiden Dateien sicherlich problemlos zu versionieren sind macht die content.xml genau das Problem, was wir auch schon in der *camaign-File haben: Es fasst mehrere separate Code-Teile in einer Datei zusammen. Wenn wir sauberen Code versionieren wollen reicht es also nicht "nur" die Kampagne zu entpacken. Wir werden wohl oder übel auch noch eine XML-Datei aufteilen und wieder zusammenstöpseln müssen.

Dazu sollten wir uns mal überlegen wie wir das am besten hinbekommen und wie das am sinnvollsten ist.

Mein naiver Erstvorschlag diesbezüglich wäre jetzt einfach mal folgendes: Wir schreiben ein "kleines" Skript, welches wir manuell ausführen können. Dieses sollen dann im 1. Schritt eine *.campaing-Datei entzippen, ODER aus der Ordnerstruktur eine *.zip-Datei zusammenstellen für den Release. Selbiges könnten wir auch für die content.XML machen. Wobei es hier schon schwieriger wird, da die einzelnen Makros Tokens zugeordnet sind. Außerdem gibt es auch auf dem Kampagnen- und dem SL-Fenster Makros, welche jedoch sehr klein sind und meist nur einen Aufruf auf ein Lib:Token-Makro machen. Die könnten wir also erstmal ignorieren. Man könnte also ein Skript schreiben, welches die Makros der einzelnen Lib:Tokens als separate Datei extrahiert. Natürlich bräuchten wir dann auch hier wieder ein Gegenstück um die Makros wieder zu einem Lib:Token zusammenzubauen. Da alles XML ist sollte das nicht sonderlich schwer, aber u.U. etwas aufwändig sein.
Ebenfalls zu beachten wäre der Workflow beim Coden. Aktuell schreibe ich den Code in der offenen Maptool-Anwendung. Wobei ich sagen muss dass ich damit nicht unbedingt zufrieden bin. Zwar kann ich den Code direkt testen, jedoch lässt der Editor in maptool sehr zu wünschen übrig. Wenn ich mehrere Skripte gleichzeitig öffne habe ich z.B. mehrere Fenster offen, die ich alle einzeln verschieben und anordnen muss. Wenn es möglich wäre in einem externen Editor zu arbeiten könnte das zwar Abhilfe schaffen aber es würden auch neue Probleme auftreten. Z.B. müsste man jedes Mal die *.cmpaign-Datei neu zusammenstöpseln um den Code zu testen. Für eine anständige Versionierung vertrete ich aber die Meinung dass separate Dateien für jedes Makro Sinn machen.

Ich hoffe ich habe damit etwas Klarheit geschaffen wie der aktuelle Stand ist und was möglich wäre. Lass gerne mal deine Meinung und Vorschläge hören was du für sinnvoll erachten würdest.

**Comment by Lector** [comment]
Ich habe mich mal bemüht die content.xml etwas näher zu analysieren. Ich glaube ich habe jetzt genug Wissen zusammen um zu beschreiben wie man vorgehen könnte:

1. *.campaign entzippen. Die aktuelle MT-Version hat übrigens noch ein par mehr Unterordner als in meinem letzten Beitrag beschrieben.
2. Makros aus content.xml extrahieren. Dazu folgende Tags iterieren: **<net.rptools.maptool.model.MacroButtonProperties>**
Innerhalb dieser Tags gibt es ein Tag **<saveLocation>...</saveLocation>** Darin steht entweder **CampaignPanel** oder **Token**

Wenn Token: Den Innhalt von **<command>** als Makro-Datei auslagern. Dabei XML-Sequenzen beachten. Die ausgelagerte Datei sollte so heißen wie unter **<label>** angegeben ist.

Wenn CampaignPanel: Vorsicht! Auslagern wird erstmal ein kleines Problem machen da <label> hier HTML-Icons sind die sich nicht als Dateinamen eignen. Natürlich könnte man auch **<macroUUID>** als Dateinamen nehmen da dieser eindeutig ist. Da hier jedoch nur winzige Makros liegen die oft nur auf ein Token-Makro verweisen mein Vorschlag: Erstmal so lassen.

Ich schlage vor das **<command>...</command>**-Tag zu ersetzen durch so etwas wie **<command src="relativer Dateipfad"/>** in der Hoffnung dass Maptool diese Syntax nicht irgendwie intern nutzt und wir das ganze wieder eindeutig rückführen können.

**Comment by Naxos84** [comment]
Ich habe mal angefangen ein CLI Tool mit Javascript zu schreiben.

Konnte meine Testkampagne splitten und wieder mergen und die hat danach auch funktioniert :D

https://github.com/Naxos84/maptool-campaign-tool.git

um das laufen zu lassen braucht es NodeJS
Dann einfach `npm run build` und `node ./dist/index.js split ./Path/To/Campain.cmpgn ./Path/To/Output` bzw. `node ./dist/index.js merge ./Path/ToSplittedFolder ./Path/To/Result/Campaign.cmpgn` machen.

**Reply by Lector** [reply]
Finde ich gut. Ich freue mich auf deinen konstruktiven Beitrag.

1. Frage: Splittet das auch schon die Einzelskripte?
2. Hast du es schon mit der aktuellen DSA5-Kampagne getestet?

Wenn ich dein ok bekomme würde ich meinen Stand mal aufsplitten und hier auf die Einzeldateien umstellen. Was ich hier auch noch gerne hätte wäre dann eine Batch-Datei welche split und merge anhand eines relativen Dateipfades durchführt.

**Reply by Naxos84** [reply]
Zu 1. 
Nein tut es noch nicht.
Zu 2.
Momentan ist es ein Node Skript. Heißt man braucht nodejs um es auszuführen.
Selbstverständlich kann man mit relativen Pfaden arbeiten.
Man wird Quelle und Ziel angeben müssen.

Momentan bin ich dabei vorzubereiten, dass ich besser an alle Stellen ran komme (ist ja aktuell eine riesen datei).
Komme nur langsam vorran, weil wenig Zeit.

Mit der dsa5 Kampagne setze ich das ganze um, denn die ist ja der Ausgangspunkt, warum es das Tool gibt.
Sobald ich etwas habe, was du testen kannst, gebe ich natürlich bescheid.
